### PR TITLE
Fix hierarchy integration from NukeStudio/Hiero.

### PR DESCRIPTION
**Pull request by tokejepsen**
Publishing a review (or uploading) when the hierachy was not already created, would result in the Ftrack component not being created.

### CODE REVIEW

File: client/ayon_ftrack/plugins/publish/integrate_hierarchy_ftrack.py [hunk 1]
```diff
@@ -273,7 +273,7 @@ def import_to_ftrack(
 
             # Create entity if not exists
             if entity is None:
-                folder_type = entity_data["folder_type"]
+                folder_type = entity_data["folder_type"].capitalize()
```
Comment (iLLiCiTiT) [inline client/ayon_ftrack/plugins/publish/integrate_hierarchy_ftrack.py:276]: Why `capitalize`?

File: client/ayon_ftrack/plugins/publish/integrate_hierarchy_ftrack.py [hunk 2]
```diff
@@ -287,6 +287,13 @@ def import_to_ftrack(
             instances = []
             for instance in context:
                 instance_folder_path = instance.data.get("folderPath")
+
+                # Ensure project name is the first entry in the folder path.
+                if not instance_folder_path.startswith("/" + project_name):
```
Comment (iLLiCiTiT) [inline client/ayon_ftrack/plugins/publish/integrate_hierarchy_ftrack.py:292]: But `entity_path` should not contain project name. If it does, then we should make sure it does not...

Comment (iLLiCiTiT) [reply client/ayon_ftrack/plugins/publish/integrate_hierarchy_ftrack.py:292]: Rather change [this line](https://github.com/ynput/ayon-ftrack/pull/88/files#diff-62a6c74c35c702c33b39c835c1846ecaf770a1735df0b12fd6e5a0a4b88f8509R283) to.
```python
            if entity_type.lower() == "project":
                entity_path = ""
            else:
                entity_path = "{}/{}".format(parent_path, entity_name)
```

File: client/ayon_ftrack/plugins/publish/integrate_hierarchy_ftrack.py (hunk 1 above)
Comment (tokejepsen) [reply client/ayon_ftrack/plugins/publish/integrate_hierarchy_ftrack.py:276]: Ok, will investigate a more flexible solution. For context, this is the error:

```
Traceback (most recent call last):
  File "C:\Users\tokejepsen\AppData\Local\Ynput\AYON\dependency_packages\ayon_2402141620_windows.zip\dependencies\pyblish\plugin.py", line 527, in __explicit_process
    runner(*args)
  File "C:\Users\tokejepsen\ayon-ftrack\client\ayon_ftrack\plugins\publish\integrate_hierarchy_ftrack.py", line 75, in process
  File "C:\Users\tokejepsen\ayon-ftrack\client\ayon_ftrack\plugins\publish\integrate_hierarchy_ftrack.py", line 277, in import_to_ftrack
  File "C:\Users\tokejepsen\AppData\Local\Ynput\AYON\dependency_packages\ayon_2402141620_windows.zip\dependencies\ftrack_api\session.py", line 643, in create
    entity = self._create(entity_type, data, reconstructing=reconstructing)
  File "C:\Users\tokejepsen\AppData\Local\Ynput\AYON\dependency_packages\ayon_2402141620_windows.zip\dependencies\ftrack_api\session.py", line 652, in _create
    raise ftrack_api.exception.UnrecognisedEntityTypeError(entity_type)
ftrack_api.exception.UnrecognisedEntityTypeError: Entity type "b'sequence'" not recognised.
```

File: client/ayon_ftrack/plugins/publish/integrate_hierarchy_ftrack.py (hunk 2 above)
Comment (tokejepsen) [reply client/ayon_ftrack/plugins/publish/integrate_hierarchy_ftrack.py:292]: ok, will investigate where the problem is.

Comment (iLLiCiTiT) [reply client/ayon_ftrack/plugins/publish/integrate_hierarchy_ftrack.py:292]: The suggested change should fix any issues you hit.

Comment (tokejepsen) [reply client/ayon_ftrack/plugins/publish/integrate_hierarchy_ftrack.py:292]: > Rather change [this line](https://github.com/ynput/ayon-ftrack/pull/88/files#diff-62a6c74c35c702c33b39c835c1846ecaf770a1735df0b12fd6e5a0a4b88f8509R283) to.
> 
> ```python
>             if entity_type.lower() == "project":
>                 entity_path = ""
>             else:
>                 entity_path = "{}/{}".format(parent_path, entity_name)
> ```

Which line? The link goes a diff.

File: client/ayon_ftrack/plugins/publish/integrate_hierarchy_ftrack.py (hunk 1 above)
Comment (iLLiCiTiT) [reply client/ayon_ftrack/plugins/publish/integrate_hierarchy_ftrack.py:276]:
```python
ft_project_schema = ft_project["project_schema"]
object_types_by_low_name = {
    object_type["name"].lower(): object_type["name"].replace(" ", "")
    for object_type in ft_project_schema["object_types"]
}
```

You can then use
```python
folder_type_low = entity_data["folder_type"].lower()
entity_type = object_types_by_low_name [folder_type_low]
entity = session.create(entity_type, {
    ...
})
```

File: client/ayon_ftrack/plugins/publish/integrate_hierarchy_ftrack.py [hunk 3]
```diff
@@ -247,6 +247,11 @@ def import_to_ftrack(
             str(ft_task_statuses)
         )
 
+        object_types_by_lower_name = {
+            x["name"].lower(): x
+            for x in ft_project["project_schema"]["object_types"]
```
Comment (iLLiCiTiT) [inline client/ayon_ftrack/plugins/publish/integrate_hierarchy_ftrack.py:252]:
```suggestion
            obj_type["name"].lower(): obj_type
            for obj_type in ft_project["project_schema"]["object_types"]
```

File: client/ayon_ftrack/plugins/publish/integrate_hierarchy_ftrack.py [hunk 4]
```diff
@@ -273,7 +278,11 @@ def import_to_ftrack(
 
             # Create entity if not exists
             if entity is None:
-                folder_type = entity_data["folder_type"]
+                # Sanitize against case sensitive folder types.
+                folder_type = object_types_by_lower_name[
+                    entity_data["folder_type"].lower()
+                ]["name"]
```
Comment (iLLiCiTiT) [inline client/ayon_ftrack/plugins/publish/integrate_hierarchy_ftrack.py:284]:
```suggestion
                folder_type_low = entity_data["folder_type"].lower()
                object_type = object_types_by_lower_name[folder_type_low]
                entity_type = object_type["name"].replace(" ", "")
```
Requires to change 286 to
`entity = session.create(entity_type, {`

File: client/ayon_ftrack/plugins/publish/integrate_hierarchy_ftrack.py (hunk 2 above)
Comment (iLLiCiTiT) [reply client/ayon_ftrack/plugins/publish/integrate_hierarchy_ftrack.py]: Currently on line 283 there is
```
entity_path = "{}/{}".format(parent_path, entity_name)
```

**Pull request merged**